test/e2e_node/containers_lifecycle_test.go

```diff
@@ -603,3 +605,282 @@ func preparePod(pod *v1.Pod) {
 		},
 	}
 }
+
+var _ = SIGDescribe("[NodeAlphaFeature:SidecarContainers] Containers Lifecycle ", func() {
+	f := framework.NewDefaultFramework("containers-lifecycle-test")
+	f.NamespacePodSecurityEnforceLevel = admissionapi.LevelPrivileged
+
+	ginkgo.When("using a Pod with restartPolicy=Never, three init container and two sidecars", ginkgo.Ordered, func() {
+
+		init1 := "init-1"
+		sidecar1 := "sidecar-1"
+		init2 := "init-2"
+		sidecar2 := "sidecar-2"
+		init3 := "init-3"
+		regular1 := "regular-1"
+
+		podSpec := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "sidecar-containers-start-serially",
+			},
+			Spec: v1.PodSpec{
+				RestartPolicy: v1.RestartPolicyNever,
+				InitContainers: []v1.Container{
+					{
+						Name:  init1,
+						Image: busyboxImage,
+						Command: ExecCommand(init1, execCommand{
+							Delay:    1,
+							ExitCode: 0,
+						}),
+					},
+					{
+						Name:  sidecar1,
+						Image: busyboxImage,
+						Command: ExecCommand(sidecar1, execCommand{
+							Delay:    600, // replacing to 1 will make test fail
```

Comment:
@mrunalp this is what I just told you about. If this will be `1`, i.e. sidecar exits before the regular containers started, it will keep waiting for sidecar to start over and over.

Reply:
I think it is because of shouldRestartContainer function - we need extra piece of information

Reply:
> @mrunalp this is what I just told you about. If this will be `1`, i.e. sidecar exits before the regular containers started, it will keep waiting for sidecar to start over and over.

addressed this issue and confirmed with test cases https://github.com/SergeyKanzhelev/kubernetes/pull/11#discussion_r1135306357